runtime/domain.c

```diff
@@ -845,7 +843,7 @@ static void* domain_thread_func(void* v)
     domain_terminate();
     /* Joining domains will lock/unlock the terminate_mutex so this unlock will
        release them if any domains are waiting. */
-    sync_mutex_unlock(terminate_mutex);
+    caml_ml_mutex_unlock(ml_values->mutex);
```

Comment:
Thanks a lot for the detailed analysis and explanations!  My mental picture was less detailed, as I viewed the Custom heap block and the sync_mutex pointer it contains as "equally valid" (i.e. both are valid or both were deallocated by the GC).  But that's not true if a minor GC is concurrently moving heap blocks, indeed.

I'll merge your comment in this PR.  I'll look into unlocking in `domain_terminate`.

Reply:
I'm a bit worried that this will become false if we someday reimplement heap compaction.
I don't think we should change this code to prepare for that, but maybe add a comment so that whoever implements compaction has a chance to learn about the problem?

Reply:
Tentative implementation of unlocking in `domain_terminate` implemented in 2db69380a .  I tried to release the termination lock "as late as possible".

Reply:
I think https://github.com/ocaml/ocaml/commit/2db69380aa9193c753eddd2fc60ef11a8137453d is sound. While holding the `all_domains_lock`, no new STW section can start (as in the comment around L1361) and so no minor or major garbage collections can initiate while this domain drops itself from the STW set. 

The niggle is the domain join completes but the terminating domain continues with runtime cleanup; however the terminating domain never runs OCaml code after that point. We could decide this is fine. 

There is a third option: pull the custom sync_mutex pointer out of the OCaml block in a stack variable inside `domain_thread_func` (i.e. the variable `terminate_mutex` in the original code). This sync_mutex pointer can not be moved by the garbage collector and so liveness is the only concern which is handled by the global root. 

Advantages: 
- we unblock joining threads at the very end (maybe OCaml based timing of domain terminate would be more accurate? maybe tighter resource usage?)

Disadvantages: 
- needing to provide a sync interface both for the ML level and operating on the pointers in the cusom blocks 
- the more complex soundness argument for the code might be a liability

I'm leaning in favour of the simplicity of https://github.com/ocaml/ocaml/commit/2db69380aa9193c753eddd2fc60ef11a8137453d  until there is a use case problem with the cleanup code and join.

Reply:
> until there is a use case problem with the cleanup code and join.

I can think of a possible problem: in "cleanup at exit" mode, if other domains or threads are still running, memory can be freed "under their feet" and cause crashes.  So the advice is to join all threads or domains before the end of the initial domain.  But here some domains may still be running cleanup code after the join, which could interfere with the "cleanup at exit".

I'll give your third option a try.  The <caml/sync.h> interface will need a system-dependent definition of `sync_mutex`, but that should be doable.